Add qml.density_matrix QNode return with partial trace capabilities

rmoyard opened the pull request:
**Context:**

Pull request related to the issue #867 , the aim is to add the possibility of returning reduced density matrix (possibly mixed state) given a set of wires to be traced out.

**Description of the Change:**
By using the 'default.qubit' device, the user has now the possibility to return for the reduced density matrix (given a set of wires). For the device 'default.qubit' it constructs the reduced density matrix directly from the state vector and the list of wires to be traced out.

**Example**
The following code
```
qml.enable_tape()

dev = qml.device("default.qubit", wires=2)

@qml.qnode(dev)
def circuit(x):
qml.PauliY(wires=0)
    qml.Hadamard(wires=1)
    return qml.density_matrix(wires=[1])  # wire 0 is traced out

print(circuit())

```
gives the reduced density matrix after tracing out 0. It is a complex matrix with trace = 1 as this is a condition for being a quantum state.
```
[[0.5+0.j 0.5+0.j]
 [0.5+0.j 0.5+0.j]]
```

It also works with the device default.mixed:
```
qml.enable_tape()

dev = qml.device("default.mixed", wires=2)

@qml.qnode(dev)
def circuit(x):
qml.PauliY(wires=0)
    qml.Hadamard(wires=1)
    return qml.density_matrix(wires=[1])  # wire 0 is traced out

print(circuit())

trbromley commented:
Thanks @rmoyard! I'd be happy to take a look, just let us know when the PR is ready for some feedback.

One question I had from your code block above:
> `return qml.density_matrix(wires=[0])  # wire 0 is traced out`
> gives the reduced density matrix after tracing out 0

Just to clarify, the return would be:
- The state of wire 0 with wire 1 traced out (what I expect)
- The state of wire 1 with wire 0 traced out?

Thanks!

rmoyard commented:
@trbromley Thank you for the question, indeed I had misunderstood what were the wires for inputs. It is updated in the new version!

@josh146 @antalszava I added the density matrix with partial trace to the mixed state device and also more tests for both default.qubit and default.mixed. The PR fails the test `Tests / device-tests (default.mixed, true, 1000) (pull_request)` because I added the capability 'return state' to the default.mixed. Should I remove the capability and the tests related to default.mixed in the file `test_tape_measure`? Except this problem, the PR is ready for review!

josh146 commented:
Thank you @rmoyard! We will go through and review today

josh146 commented on pennylane/devices/default_mixed.py:
```diff
@@ -118,13 +119,55 @@ def _create_basis_state(self, index):
+        traced_wires = Wires(traced_system)
+
+        # Trace first subsystem by applying kraus operators of the partial trace
+        tr_op = np.asarray(np.eye(2).reshape(1, 4), dtype=complex)
```
```suggestion
        tr_op = np.asarray(np.eye(2).reshape(1, 4), dtype=self.C_DTYPE))
```

josh146 commented on pennylane/devices/default_mixed.py:
```diff
@@ -118,13 +119,55 @@ def _create_basis_state(self, index):
+        tr_op = np.asarray(np.eye(2).reshape(1, 4), dtype=complex)
+        tr_op = self._reshape(tr_op, (2, 1, 2))
+
+        self._apply_channel(tr_op, traced_wires[0])
+
+        # Trace next subsystem by applying kraus operators of the partial trace
+        for traced_wire in traced_wires[1:]:
+            self._apply_channel(tr_op, traced_wire)
```
This is a clever way of implementing it!

However, I wonder if there is a more efficient method that avoids a slow Python for loop. For example, by using `np.einsum` or `np.tensordot` to perform the partial trace of _all_ non-specified wires simultaneously.

rmoyard replied:
The number of Kraus operators for the partial trace is following 2^(#number all non-specified wires) and we need to construct all of them by using `np.tensordot`. I think it leads to a bigger number of operations overall compared to applying successively the Kraus operator for one system and the code will be more complicated to understand.

josh146 commented on pennylane/devices/default_mixed.py:
```diff
@@ -181,9 +224,17 @@ def _apply_channel(self, kraus, wires):
-        kraus_shape = [len(kraus)] + [2] * num_ch_wires * 2
-        kraus = self._cast(self._reshape(kraus, kraus_shape), dtype=self.C_DTYPE)
-        kraus_dagger = self._cast(self._reshape(kraus_dagger, kraus_shape), dtype=self.C_DTYPE)
+        if kraus[0].shape[0] == kraus[0].shape[1]:
+            kraus_shape = [len(kraus)] + [2] * num_ch_wires * 2
+            kraus = self._cast(self._reshape(kraus, kraus_shape), dtype=self.C_DTYPE)
+            kraus_dagger = self._cast(self._reshape(kraus_dagger, kraus_shape), dtype=self.C_DTYPE)
+        elif (kraus[0].shape == (1, 2)) and (num_ch_wires == 1):
+            kraus_shape = [len(kraus)] + list(kraus[0].shape)
+            kraus = self._cast(self._reshape(kraus, kraus_shape), dtype=self.C_DTYPE)
+            kraus_dagger_shape = [len(kraus)] + list(kraus[0].shape)[::-1] * num_ch_wires
+            kraus_dagger = self._cast(
+                self._reshape(kraus_dagger, kraus_dagger_shape), dtype=self.C_DTYPE
+            )
```
Is this change primarily to support passing kraus operators of the partial trace? I might tag @ixfoduap for verification 🙂

rmoyard replied:
It only accepted Kraus operators of size nxn (e.g. 2x2, 4x4) and the one for partial traces have shape (1x2). I was unable to change the code for accepting every type of shapes, therefore I added the needed case. But it would be great to change it.

josh146 commented on pennylane/devices/default_qubit.py:
```diff
@@ -412,6 +412,35 @@ def _create_basis_state(self, index):
         return self._flatten(self._pre_rotated_state)
 
+    def density_matrix(self, wires=None):
+        """Return the reduced density matrix given a set of wires to be traced out.
```
Just double checking, but should this instead be "Return the reduced density matrix of a given set of wires"?

rmoyard replied:
Yes I've forgotten to change it.

josh146 commented on tests/tape/test_tape_measure.py:
```diff
@@ -501,3 +512,360 @@ def func():
+class TestDensityMatrix:
+    """Tests for the density matrix function"""
+
+    @pytest.mark.parametrize("wires", range(2, 5))
+    def test_density_matrix_shape_and_dtype_qubit(self, wires):
+        """Test that the density matrix is of correct size and dtype for a
+        trivial circuit"""
+
+        dev = qml.device("default.qubit", wires=wires)
+
+        @qnode(dev)
+        def circuit():
+            return density_matrix([0])
+
+        state_val = circuit()
+
+        assert state_val.shape == (2, 2)
+        assert state_val.dtype == np.complex128
+
+    @pytest.mark.parametrize("wires", range(2, 5))
+    def test_density_matrix_shape_and_dtype_mixed(self, wires):
+        """Test that the density matrix is of correct size and dtype for a
+        trivial circuit"""
+
+        dev = qml.device("default.mixed", wires=wires)
+
+        @qnode(dev)
+        def circuit():
+            return density_matrix([0])
+
+        state_val = circuit()
+
+        assert state_val.shape == (2, 2)
+        assert state_val.dtype == np.complex128
```
You could combine these two tests into a single test, using

```python
@pytest.mark.parametrize("wires", range(2, 5))
@pytest.mark.parametrize("dev_name", ["default.qubit", "default.mixed"])
def test_density_matrix_shape_and_dtype_mixed(self, dev_name, wires):
```

josh146 commented on tests/tape/test_tape_measure.py:
```diff
@@ -501,3 +512,360 @@ def func():
+
+    def test_correct_density_matrix_product_state_first_qubit(self):
+        """Test that the correct density matrix  is returned when
+        trace out a product state"""
```
```suggestion
        tracing out a product state"""
```

josh146 commented on tests/tape/test_tape_measure.py:
```diff
@@ -501,3 +512,360 @@ def func():
+
+    def test_correct_density_matrix_product_state_first_mixed(self):
+        """Test that the correct density matrix  is returned when
+        trace out a product state"""
```
```suggestion
        tracing out a product state"""
```

josh146 commented on tests/tape/test_tape_measure.py:
```diff
@@ -501,3 +512,360 @@ def func():
+
+    def test_correct_density_matrix_product_state_second_qubit(self):
+        """Test that the correct density matrix  is returned when
+        trace out a product state"""
```
```suggestion
        tracing out a product state"""
```

josh146 replied:
This docstring is the same as the previous test - would be good to expand to detail how/why this test differs

josh146 commented on tests/tape/test_tape_measure.py:
```diff
@@ -501,3 +512,360 @@ def func():
+
+    def test_correct_density_matrix_product_state_second_mixed(self):
+        """Test that the correct density matrix  is returned when
+        trace out a product state"""
```
```suggestion
        tracing out a product state"""
```

josh146 commented on tests/tape/test_tape_measure.py:
```diff
@@ -501,3 +512,360 @@ def func():
+            np.array([[0.5 + 0.0j, 0.5 + 0.0j], [0.5 + 0.0j, 0.5 + 0.0j]]), density_second
+        )
+
+    def test_correct_density_matrix_all_wires_qubit(self):
+        """Test that the correct density matrix  is returned when
+        all wires are given"""
```
A nice test to ensure that the partial trace logic is skipped is to use the `mocker` fixture. In particular, [`mocker.spy`](https://github.com/pytest-dev/pytest-mock/#spy) allows you to 'spy' on a function/method call, and assert whether it has been called or not.

josh146 commented on tests/tape/test_tape_measure.py:
```diff
@@ -501,3 +512,360 @@ def func():
+            qml.Hadamard(wires=0)
+            return density_matrix(0), expval(qml.PauliZ(1))
+
+        with pytest.raises(
+            QuantumFunctionError,
+            match="The state respectively density matrix"
+            " cannot be returned in combination"
+            " with other return types",
+        ):
+            func()
```
Nice!

josh146 commented on tests/tape/test_tape_measure.py:
```diff
@@ -501,3 +512,360 @@ def func():
+                qml.CNOT(wires=[wires[i], wires[i + 1]])
+            return density_matrix(0)
+
+        with pytest.raises(QuantumFunctionError, match="custom wire labels"):
+            func()
```
@trbromley, I suppose the restriction on custom wire labels can be lifted here, since the `qml.density_matrix` function should always receive wire arguments?

josh146 replied:
@trbromley just tagging again in case you missed this. Or was this restriction due to Cirq, so should remain for density matrices?

trbromley replied:
Sorry I totally missed this! Yes the main worry with `qml.state()` was compatibility with Cirq and any other device that can have a custom wire order. So I think it also makes sense here.

josh146 commented on tests/tape/test_tape_measure.py:
```diff
@@ -501,3 +512,360 @@ def func():
+            return density_matrix(0)
+
+        with pytest.raises(QuantumFunctionError, match="custom wire labels"):
+            func()
```
Awesome tests @rmoyard 💪

rmoyard replied:
@josh146 Thank you for the review, I will go through the different points today!

josh146 reviewed (commented):
Thanks @rmoyard! This is looking pretty great, and great test coverage as well.

I suppose my biggest question relates to `default.mixed`; is there likely a more efficient way we could implement the partial trace, that avoids a for loop?

rmoyard commented on pennylane/devices/default_mixed.py:
```diff
@@ -118,13 +119,55 @@ def _create_basis_state(self, index):
+    def capabilities(cls):
+        capabilities = super().capabilities().copy()
+        capabilities.update(
+            model="mixed",
```
Ok, then I guess it is not necessary to delete the capabilities in order to successfully pass the `Tests / device-tests (default.mixed, true, 1000) (pull_request)`?

josh146 replied:
Yep, I believe so. Maybe just `return_state` needs to be updated here, not `model`.

rmoyard replied:
Then I also need to change `return_state` capability to `True` directly in each test ?

josh146 replied:
No, I mean instead that we can keep this here, just remove the `model="mixed"` line 🙂 

So it should look like this:
```python
    @classmethod
    def capabilities(cls):
        capabilities = super().capabilities().copy()
        capabilities.update(returns_state=True)
```

rmoyard replied:
Thank you very much!

rmoyard commented on pennylane/devices/default_mixed.py:
```diff
@@ -118,13 +119,55 @@ def _create_basis_state(self, index):
+        if wires == self.wires:
+            return self.state
+
+        wires = list(wires.labels)
```
It is a tuple, the transformation is unnecessary for what we need here you are right!

rmoyard commented on pennylane/_qubit_device.py:
```diff
@@ -373,6 +376,9 @@ def access_state(self):
             raise QuantumFunctionError("The state is not available in the current device")
 
+        if wires:
+            return getattr(self, "density_matrix", None)(wires)
```
It is right, I changed it.

josh146 commented:
By the way @rmoyard, we're having some trouble with our coverage CI Codecov lately! If it seems to be giving incorrect results, you could run `make coverage` locally to ensure that your branch matches master 🙂

rmoyard commented:
@josh146 I changed and added some tests for better coverage. The only part of the code that I could not access was `raise NotImplementedError` for the `density_matrix` function in `_qubit_device.py`. The overall coverage should be fine from what I have seen by using `make coverage`. 

If there is anything else I could improve, do not hesitate!

antalszava commented on .github/CHANGELOG.md:
````diff
@@ -9,23 +9,47 @@
-  [(#871)](https://github.com/PennyLaneAI/pennylane/pull/871)
-
-  ```python
-  dev = qml.device("default.qubit", wires=1)
+[(#871)](https://github.com/PennyLaneAI/pennylane/pull/871)
 
-  @qml.qnode(dev)
-  def circuit():
+    ```python
+    dev = qml.device("default.qubit", wires=1)
+    
+    @qml.qnode(dev)
+    def circuit():
       qml.SX(wires=[0])
       return qml.expval(qml.PauliZ(wires=[0]))
-  ```
+    ```
````
The whitespace was correct here, might need a revert (e.g., the link `  [(#871)](https://github.com/PennyLaneAI/pennylane/pull/871)` was correct with the leading whitespace, also the indentation for the code snippet)

antalszava commented on .github/CHANGELOG.md:
```diff
@@ -9,23 +9,47 @@
-  `tape.get_resources()` returns a dictionary with a list of the constituent operations and the
-  number of times they were run. Similarly, `tape.get_depth()` computes the circuit depth.
-  [(#862)](https://github.com/PennyLaneAI/pennylane/pull/862)
+* Add qml.density_matrix QNode return with partial trace capabilities.
```
```suggestion
* Add the `qml.density_matrix` QNode return with partial trace capabilities.
```

josh146 commented on tests/tape/test_tape_measure.py:
```diff
@@ -321,10 +321,10 @@ def func():
+                match="The state or density matrix"
+                      " cannot be returned in combination"
+                      " with other return types",
         ):
```
The whitespace change looks a bit weird here 🤔

josh146 reviewed (commented):
Looks great @rmoyard 🙂 I'm submitting my review slightly early, since I can see @antalszava leaving comments at the same time!

antalszava commented on pennylane/tape/measure.py:
```diff
@@ -375,3 +375,54 @@ def circuit():
+    array([[0.+0.j 0.+0.j]
+        [0.+0.j 1.+0.j]])
+
+    The returned matrix is the reduced density matrix, where the system 1 is traced out.
```
```suggestion
    The returned matrix is the reduced density matrix, where system 1 is traced out.
```

antalszava commented on pennylane/devices/default_mixed.py:
```diff
@@ -181,9 +221,17 @@ def _apply_channel(self, kraus, wires):
+        elif (kraus[0].shape == (1, 2)) and (num_ch_wires == 1):
+            kraus_shape = [len(kraus)] + list(kraus[0].shape)
+            kraus = self._cast(self._reshape(kraus, kraus_shape), dtype=self.C_DTYPE)
+            kraus_dagger_shape = [len(kraus)] + list(kraus[0].shape)[::-1] * num_ch_wires
```
As per the condition, `(num_ch_wires == 1) == True` stands
```suggestion
            kraus_dagger_shape = [len(kraus)] + list(kraus[0].shape)[::-1]
```

antalszava commented on pennylane/_qubit_device.py:
```diff
@@ -94,6 +94,7 @@ class QubitDevice(Device):
     _outer = staticmethod(np.outer)
     _diag = staticmethod(np.diag)
     _real = staticmethod(np.real)
+    _delete = staticmethod(np.delete)
```
Was this used eventually?

rmoyard replied:
No it was for a first version of the partial trace, I'll remove it.

antalszava commented on tests/tape/test_tape_measure.py:
```diff
@@ -501,3 +512,235 @@ def func():
             func()
+
+
+class TestDensityMatrix:
```
Awesome tests! :100: :)

antalszava reviewed (commented):
Hi @rmoyard, this is looking great! :blush: My main questions is related to the default argument of `wires` when returning the density matrix representation of a state. Other than that the rest of my thoughts/questions were minor.

rmoyard commented on pennylane/devices/default_mixed.py:
```diff
@@ -118,13 +119,52 @@ def _create_basis_state(self, index):
+        traced_wires = Wires(traced_system)
+
+        # Trace first subsystem by applying kraus operators of the partial trace
+        tr_op = self._cast(np.eye(2).reshape(1, 4), dtype=self.C_DTYPE)
+        tr_op = self._reshape(tr_op, (2, 1, 2))
```
It can indeed be simplified, I'll remove it.

rmoyard commented on pennylane/devices/default_qubit.py:
```diff
@@ -412,6 +412,35 @@ def _create_basis_state(self, index):
+        dim = self.num_wires
+        state = self._pre_rotated_state
+
+        # Return the reduced density matrix by using numpy tensor product
```
I changed the comment by specifying that the density matrix is for the complete system. I added a second comment to emphasize that it is the density matrix for the reduced system.

rmoyard commented on pennylane/devices/default_mixed.py:
```diff
@@ -118,13 +119,52 @@ def _create_basis_state(self, index):
         # User obtains state as a matrix
         return self._reshape(self._pre_rotated_state, (dim, dim))
 
+    def density_matrix(self, wires=None):
```
You are right it should not be equal to `None` because the `density_matrix` is called if wires were given `if wires`. This was done in order to differentiate the function `state() = MeasurementProcess(State)` and `density_matrix(wires) = MeasurementProcess(State, wires)`. As we are dealing with quantum system, trace out all subsystems will always return 1 and therefore I don't think it is useful to implement it. Consequently I'll just remove the default argument.

rmoyard commented:
@josh146 @antalszava Thank you for the useful comments! I answered the different questions and updated the PR according to the comments.

josh146 commented on .github/CHANGELOG.md:
```diff
@@ -9,22 +9,36 @@
-  `tape.get_resources()` returns a dictionary with a list of the constituent operations and the
-  number of times they were run. Similarly, `tape.get_depth()` computes the circuit depth.
-  [(#862)](https://github.com/PennyLaneAI/pennylane/pull/862)
+* Adds the `qml.density_matrix` QNode return with partial trace capabilities.
+  [(#878)](https://github.com/PennyLaneAI/pennylane/pull/878)
+  The density matrix over the provided wires is returned, with all other subsystems traced out.
```
```suggestion
* Adds the `qml.density_matrix` QNode return with partial trace capabilities.
  [(#878)](https://github.com/PennyLaneAI/pennylane/pull/878)

  The density matrix over the provided wires is returned, with all other subsystems traced out.
```

josh146 reviewed (approved):
Thanks @rmoyard, just a slight formatting change in the changelog, other than that, fantastic 🎉

antalszava reviewed (approved):
Looks good to me, great addition!  :) :confetti_ball: Thanks for considering the suggestions and the update!

trbromley commented on pennylane/_qubit_device.py:
```diff
@@ -356,14 +358,14 @@ def statistics(self, observables):
-    def access_state(self):
+    def access_state(self, wires=None):
         """Check that the device has access to an internal state and return it if available.
 
```
Might be good to add the `wires` argument to the docstring here.

trbromley commented on pennylane/devices/default_mixed.py:
```diff
@@ -181,9 +221,17 @@ def _apply_channel(self, kraus, wires):
-        kraus_shape = [len(kraus)] + [2] * num_ch_wires * 2
-        kraus = self._cast(self._reshape(kraus, kraus_shape), dtype=self.C_DTYPE)
-        kraus_dagger = self._cast(self._reshape(kraus_dagger, kraus_shape), dtype=self.C_DTYPE)
+        if kraus[0].shape[0] == kraus[0].shape[1]:
```
Might be good to add a comment to motivate this change

trbromley commented on pennylane/devices/default_qubit.py:
```diff
@@ -412,6 +412,36 @@ def _create_basis_state(self, index):
+            wires (Sequence[int] or int): the wires of the subsytem.
+
+        Returns:
+            tensor[complex]: complex tensor of shape ``2^(self.num_wires-len(wires))``
```
```suggestion
            tensor[complex]: complex tensor of shape ``2 ** (self.num_wires-len(wires))``
```

trbromley commented on pennylane/devices/default_qubit.py:
```diff
@@ -412,6 +412,36 @@ def _create_basis_state(self, index):
+        """Returns the reduced density matrix of a given set of wires.
+
+        Args:
+            wires (Sequence[int] or int): the wires of the subsytem.
```
```suggestion
            wires (Wires): wires of the reduced system
```

trbromley commented on pennylane/devices/default_qubit.py:
```diff
@@ -412,6 +412,36 @@ def _create_basis_state(self, index):
         return self._flatten(self._pre_rotated_state)
 
+    def density_matrix(self, wires):
+        """Returns the reduced density matrix of a given set of wires.
+
+        Args:
+            wires (Sequence[int] or int): the wires of the subsytem.
+
+        Returns:
+            tensor[complex]: complex tensor of shape ``2^(self.num_wires-len(wires))``
+            representing the reduced density matrix.
```
Would be good to make sure it's the same as for default mixed, excepting the array size.

How about we just write the one docstring in `QubitDevice` and don't mention the output shape (which as far as I can tell is the only reason to have this repetition)

trbromley commented on tests/tape/test_tape_measure.py:
```diff
@@ -511,3 +522,235 @@ def func():
+
+    @pytest.mark.parametrize("dev_name", ["default.qubit", "default.mixed"])
+    def test_correct_density_matrix_product_state_first(self, dev_name):
+        """Test that the correct density matrix  is returned when
```
```suggestion
        """Test that the correct density matrix is returned when
```

trbromley commented on tests/tape/test_tape_measure.py:
```diff
@@ -511,3 +522,235 @@ def func():
+
+    @pytest.mark.parametrize("dev_name", ["default.qubit", "default.mixed"])
+    def test_correct_density_matrix_product_state_second(self, dev_name):
+        """Test that the correct density matrix  is returned when
```
```suggestion
        """Test that the correct density matrix is returned when
```

trbromley commented on tests/tape/test_tape_measure.py:
```diff
@@ -511,3 +522,235 @@ def func():
+
+    @pytest.mark.parametrize("dev_name", ["default.qubit", "default.mixed"])
+    def test_correct_density_matrix_all_wires(self, dev_name):
+        """Test that the correct density matrix  is returned when
```
```suggestion
        """Test that the correct density matrix is returned when
```

trbromley reviewed (commented):
Thanks @rmoyard, I'm a bit late to review so I just took a quick look.

rmoyard commented:
@trbromley Thank you for the comments, I updated the PR accordingly.

trbromley commented on pennylane/tape/measure.py:
```diff
@@ -386,3 +386,54 @@ def circuit():
+        For more details on tape mode, see :mod:`pennylane.tape`.
+
+    This function accepts no observables and instead instructs the QNode to return its density
+    matrix (respectively reduced density matrix). The ``wires`` argument gives the possibility
```
```suggestion
    matrix or reduced density matrix. The ``wires`` argument gives the possibility
```
The "respectively" part was not clear to me.

trbromley reviewed (approved):
Thanks @rmoyard, nice job! I have approved, although it might be nice to check some of my suggestions.

josh146 commented on pennylane/_qubit_device.py:
```diff
@@ -488,6 +497,16 @@ def state(self):
         """
         raise NotImplementedError
 
+    def density_matrix(self, wires):
+        """Returns the reduced density matrix prior to measurement.
+
+        .. note::
+
+            Only state vector simulators support this property. Please see the
+            plugin documentation for more details.
+        """
+        raise NotImplementedError
+
```
If it were a property, however, we wouldn't be able to pass `wires`? Perhaps we should be asking why `state` is a property and not a method 😆

trbromley replied:
Good point! Right, no need for a property then. I guess the argument for state being a property is that it doesn't have a wires argument.

rmoyard commented on pennylane/devices/default_mixed.py:
```diff
@@ -181,9 +221,19 @@ def _apply_channel(self, kraus, wires):
+            kraus = self._cast(self._reshape(kraus, kraus_shape), dtype=self.C_DTYPE)
+            kraus_dagger = self._cast(self._reshape(kraus_dagger, kraus_shape), dtype=self.C_DTYPE)
+
+        # Add the possibility to give a (1,2) shape Kraus operator
```
For the trace over a given single subsystem, the Kraus operator has the shape (1,2) and is therefore not a square matrix.

rmoyard commented on pennylane/tape/measure.py:
```diff
@@ -386,3 +386,54 @@ def circuit():
     """
     # pylint: disable=protected-access
     return MeasurementProcess(State)
+
```
Thank you for pointing this out!

rmoyard commented:
Thank you @trbromley , I followed your suggestions and updated the PR.
